### PR TITLE
Optimize item Remove operations

### LINKED ISSUE #6696: Item Remove's lack caching and lead to O(N^2) time
### Issue Description

Evaluating each `<SomeItem Remove="..."/>` performs O(N) work where N is the number of items already in `SomeItem`. This work is not cheap because it involves path normalization. `<SomeItem Remove="@(A)"/>` is O(N^2) as each item already in `SomeItem` is compared with each item in `A`. This is perf prohibitive with as few as small thousands of items and has come up in user feedback several times.

### Steps to Reproduce

```xml
<Project>
  <ItemGroup>
    <A Include="**/*.*" />
    <B Include="**/*.*" />
    <B Remove="@(A)" />
  </ItemGroup>
  <Target Name="Build" />
</Project>
```

with a few thousand files in the project directory takes seconds to evaluate.

### Data

The above project with 2000 files takes ~8 seconds to build on my machine.

### Analysis

This is where the work is happening:

```
 	Microsoft.Build.dll!Microsoft.Build.Shared.FileUtilities.GetFullPathNoThrow(string path) Line 692	C#
 	Microsoft.Build.dll!Microsoft.Build.Shared.FileUtilities.NormalizePathForComparisonNoThrow(string path, string currentDirectory) Line 739	C#
 	Microsoft.Build.dll!Microsoft.Build.Shared.FileUtilities.ComparePathsNoThrow(string first, string second, string currentDirectory, bool alwaysIgnoreCase) Line 720	C#
 	Microsoft.Build.dll!Microsoft.Build.Internal.FileSpecMatcherTester.IsMatch(string fileToMatch) Line 68	C#
 	Microsoft.Build.dll!Microsoft.Build.Evaluation.ItemSpecFragment.IsMatch(string itemToMatch) Line 415	C#
 	Microsoft.Build.dll!Microsoft.Build.Evaluation.ItemSpec<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem>.MatchesItem(Microsoft.Build.Evaluation.ProjectItem item) Line 283	C#
>	Microsoft.Build.dll!Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.RemoveOperation.SelectItems(System.Collections.Immutable.ImmutableList<Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.ItemData>.Builder listBuilder, System.Collections.Immutable.ImmutableHashSet<string> globsToIgnore) Line 60	C#
 	Microsoft.Build.dll!Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.LazyItemOperation.ApplyImpl(System.Collections.Immutable.ImmutableList<Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.ItemData>.Builder listBuilder, System.Collections.Immutable.ImmutableHashSet<string> globsToIgnore) Line 66	C#
 	Microsoft.Build.dll!Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.RemoveOperation.ApplyImpl(System.Collections.Immutable.ImmutableList<Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.ItemData>.Builder listBuilder, System.Collections.Immutable.ImmutableHashSet<string> globsToIgnore) Line 51	C#
 	Microsoft.Build.dll!Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.LazyItemOperation.Apply(System.Collections.Immutable.ImmutableList<Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.ItemData>.Builder listBuilder, System.Collections.Immutable.ImmutableHashSet<string> globsToIgnore) Line 59	C#
 	Microsoft.Build.dll!Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.MemoizedOperation.Apply(System.Collections.Immutable.ImmutableList<Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.ItemData>.Builder listBuilder, System.Collections.Immutable.ImmutableHashSet<string> globsToIgnore) Line 173	C#
 	Microsoft.Build.dll!Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.LazyItemList.ComputeItems(Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.LazyItemList lazyItemList, System.Collections.Immutable.ImmutableHashSet<string> globsToIgnore) Line 413	C#
 	Microsoft.Build.dll!Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.LazyItemList.GetItemData(System.Collections.Immutable.ImmutableHashSet<string> globsToIgnore) Line 290	C#
 	Microsoft.Build.dll!Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.GetAllItemsDeferred.AnonymousMethod__26_0(Microsoft.Build.Evaluation.LazyItemEvaluator<Microsoft.Build.Evaluation.ProjectProperty, Microsoft.Build.Evaluation.ProjectItem, Microsoft.Build.Evaluation.ProjectMetadata, Microsoft.Build.Evaluation.ProjectItemDefinition>.LazyItemList itemList) Line 489	C#
```

### Versions & Configurations

Microsoft (R) Build Engine version 17.0.0-dev-21370-01+dfd2be739 for .NET Framework

### Regression?

No.

### Attach a binlog

N/A

### Exit criteria
- Remove="someitem" is a constant-time operation (amortized).
- Remove="@(A)" scales linearly with the number of items being removed.

### PR DESCRIPTION
**Pull request by ladipro**
Fixes #6696

### Context

Removing items by referencing other item types or by multiple single-item Remove operations has bad performance characteristics because it leads to N*M comparisons, each involving path normalization of both sides. For projects with as few as small thousands of items, evaluating them becomes a bottleneck with direct impact on build time and VS responsiveness.

### Changes Made

Replaced `ImmutableList<ItemData>` with a new collection type `OrderedItemDataCollection` to be used when representing the intermediate list of items. In addition to being a list, the new collection also lazily creates a dictionary of items keyed by normalized paths, which allows for fast matching.

### Testing

Functional:
- Existing unit tests, one new unit test for a previously uncovered scenario.
- Experimental VS insertion.

Performance:
Used a benchmark project with 2000 files in a subdirectory.

Evaluation time of:
```xml
    <A Include="**/*.*" />
    <B Include="**/*.*" />
    <B Remove="@(A)" />
```
went from 7300 ms down to 180 ms.

Evaluation time of:
```xml
    <A Include="**/*.*" />    
    <A Remove='subdir\file1'/>
    <A Remove='subdir\file2'/>
    ...
    <A Remove='subdir\file2000'/>
```
went from 7900 ms down to 330 ms.

Evaluation time of a Hello World .NET Core project went got slightly faster as well (520 -> 515 ms, note that this was without NGEN so most of it is JITting).

### Notes

I opted for a rather simple implementation of `OrderedItemDataCollection` based on `ImmutableList` and `Dictionary`. It has limitations but seems to work well for the scenarios of interest. It should be possible to implement a completely custom data structure for even better results (requirements: immutable, maintains insertion order for enumeration).

### CODE REVIEW

File: src/Build/Evaluation/ItemSpec.cs [hunk 1]
```diff
@@ -415,6 +420,16 @@ public virtual bool IsMatch(string itemToMatch)
...
         }
 
+        public virtual bool IsMatchNormalized(string normalizedItemToMatch)
+        {
+            return FileMatcher.IsMatchNormalized(normalizedItemToMatch);
+        }
+
+        public virtual IEnumerable<string> GetReferencedItems()
+        {
+            yield return TextFragment;
```
Comment (Therzok) [inline src/Build/Evaluation/ItemSpec.cs:430]: Given LINQ is optimized to know about LINQ in chained operations, maybe. Unlike `yield return`, this would have support for speed optimizations like `TryGetCount`.

```suggestion
           return Enumerable.Repeat(TextFragment, 1);
```

File: src/Build/Evaluation/ItemSpec.cs [hunk 2]
```diff
@@ -446,6 +461,12 @@ public GlobFragment(string textFragment, string projectDirectory)
         {
         }
 
+        public override IEnumerable<string> GetReferencedItems()
+        {
+            // This fragment cannot efficiently enumerate its referenced items.
+            return null;
```
Comment (Therzok) [inline src/Build/Evaluation/ItemSpec.cs:467]:
```suggestion
            return Enumerable.Empty<string>();
```

File: src/Build/Evaluation/LazyItemEvaluator.cs [hunk 3]
```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
...
+                            foreach (var spec in referencedItems)
+                            {
+                                string key = FileUtilities.NormalizePathForComparisonNoThrow(spec, fragment.ProjectDirectory);
+                                if (dictionaryBuilder.TryGetValue(key, out var multiValue))
+                                {
+                                    items ??= new HashSet<I>();
+                                    foreach (I item in multiValue)
+                                    {
+                                        items.Add(item);
+                                    }
```
Comment (Therzok) [inline src/Build/Evaluation/LazyItemEvaluator.cs:400]:
```suggestion
                                    items.UnionWith(multiValue);
```

File: src/Build/Evaluation/LazyItemEvaluator.cs [hunk 4]
```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
...
+                            List<string> keysToRemove = null;
+                            foreach (var kvp in dictionaryBuilder)
+                            {
+                                if (fragment.IsMatchNormalized(kvp.Key))
+                                {
+                                    items ??= new HashSet<I>();
+                                    foreach (I item in kvp.Value)
+                                    {
+                                        items.Add(item);
+                                    }
```
Comment (Therzok) [inline src/Build/Evaluation/LazyItemEvaluator.cs:417]:
```suggestion
                                    items.UnionWith(kvp.Value);
```

File: src/Build/Evaluation/LazyItemEvaluator.cs [hunk 5]
```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
...
+
+            /// <summary>
+            /// The list of items in the collection. Defines the enumeration order.
+            /// </summary>
+            private ImmutableList<ItemData> _list;
+
+            /// <summary>
+            /// A dictionary of items keyed by their normalized value.
+            /// </summary>
+            private ImmutableDictionary<string, DictionaryValue> _dictionary;
```
Comment (Therzok) [inline src/Build/Evaluation/LazyItemEvaluator.cs:489]: ImmutableDictionary is O(logn) indexing, this might be costly?

File: src/Build/Evaluation/LazyItemEvaluator.cs [hunk 6]
```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
...
+                        {
+                            (I)_value,
+                            item
+                        };
+                    }
+                    _version = version;
+                    _value = list;
+                }
+
+                public IEnumerator<I> GetEnumerator()
```
Comment (Therzok) [inline src/Build/Evaluation/LazyItemEvaluator.cs:284]: This enumerator allocates, maybe a `struct Enumerator` can be used here to avoid this allocations on operations?

File: src/Build/Evaluation/LazyItemEvaluator.cs [hunk 7]
```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
...
+                    _version = version;
+                    _listBuilder = listBuilder;
+                    _dictionaryBuilder = dictionaryBuilder;
+                }
+
+                #region IEnumerable implementation
+
+                public IEnumerator<ItemData> GetEnumerator() => _listBuilder.GetEnumerator();
+
+                System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => _listBuilder.GetEnumerator();
```
Comment (Therzok) [inline src/Build/Evaluation/LazyItemEvaluator.cs:333]: In example, allocations can be avoided here with the ImmutableList enumerator directly:

```suggestion
                ImmutableList<ItemData>.Enumerator GetEnumerator() => _listBuilder.GetEnumerator();
                IEnumerator<ItemData> IEnumerable<T>.GetEnumerator() => _listBuilder.GetEnumerator();

                System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => _listBuilder.GetEnumerator();
```

File: src/Build/Evaluation/ItemSpec.cs (hunk 2 above)
Comment (cdmihai) [inline src/Build/Evaluation/ItemSpec.cs:467]: Should this throw NotImplementedException to protect against future use where someone hopes this works?

File: src/Build/Evaluation/LazyItemEvaluator.cs [hunk 8]
```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
...
 
                 _cache[globsToIgnore] = items;
             }
         }
 
+        /// <summary>
+        /// A collection of ItemData that maintains insertion order and internally optimizes some access patterns, e.g. bulk removal
+        /// based on normalized item values.
+        /// </summary>
+        internal sealed class OrderedItemDataCollection
```
Comment (cdmihai) [inline src/Build/Evaluation/LazyItemEvaluator.cs:237]: Would this be easy to remove in a separate file? At least yet another LazyItemEvaluator partial class

File: src/Build/Evaluation/LazyItemEvaluator.cs [hunk 9]
```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
...
+                    _value = item;
+                }
+
+                public void Add(int version, I item)
+                {
+                    if (_value is List<I> list)
+                    {
+                        if (version != _version)
+                        {
+                            list = new List<I>(list);
```
Comment (cdmihai) [inline src/Build/Evaluation/LazyItemEvaluator.cs:268]: If Add gets called a lot on the same instance on the hot path the list constructor may end up creating a lot of arrays right? Switching to ImmutableList would alleviate that.

Profiling OrchardCore would probably show if this is an issue or not.

File: src/Build/Evaluation/LazyItemEvaluator.cs [hunk 10]
```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
...
+                /// <summary>
+                /// Removes items from the collection that match the given ItemSpec.
+                /// </summary>
+                /// <remarks>
+                /// If <see cref="_dictionaryBuilder"/> does not exist yet, it is created in this method to avoid the cost of comparing each item
+                /// being removed with each item already in the collection. The dictionary is kept in sync with the <see cref="_listBuilder"/>
+                /// as long as practical. If an operation would result in too much of such work, the dictionary is simply dropped and recreated
+                /// later if/when needed.
+                /// </remarks>
+                public void RemoveMatchingItems(ItemSpec<P, I> itemSpec)
```
Comment (cdmihai) [inline src/Build/Evaluation/LazyItemEvaluator.cs:381]: Wouldn't it be better design-wise to ask the ItemSpec to compute its intersection with a given OrderedItemDataCollection via a new `IntersectWith(OrderedItemDataCollection)`?

That way you:
- prevent leaking the implementation detail that some itemspec fragments do not support enumeration 
- can neatly cache normalized values of the itemspec values without leaking that implementation detail. This would be useful if itemspec values get repeatedly normalized (for example by repeatedly calling any of the IsMatch, MatchCount, or IntersectWith methods)
- can reuse the IntersectWith operation to remove the same O(n^2) issue with the Update operation.

To support this, OrderedItemDataCollection (or the Builder) could provide a string indexer which would be the normalized key. Ideally you could even hide OrderedItemDataCollection behind some collection interface but maybe that's too much for now.

File: src/Build/Evaluation/LazyItemEvaluator.cs [hunk 11]
```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
...
+                                    keysToRemove ??= new List<string>();
+                                    keysToRemove.Add(kvp.Key);
+                                }
+                            }
+
+                            if (keysToRemove != null)
+                            {
+                                foreach (string key in keysToRemove)
+                                {
+                                    dictionaryBuilder.Remove(key);
```
Comment (Therzok) [inline src/Build/Evaluation/LazyItemEvaluator.cs:427]: ImmutableDictionary has bulk removal support In order to minimize tree mutation (node alloc for rebalance):
https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutabledictionary-2.builder.removerange?view=net-5.0

File: src/Build/Utilities/FileSpecMatchTester.cs [hunk 12]
```diff
@@ -41,31 +48,43 @@ public static FileSpecMatcherTester Parse(string currentDirectory, string fileSp
...
                 // Check file name first as it's more likely to not match.
-                string filename = Path.GetFileName(fileToMatch);
+                string filename = Path.GetFileName(normalizedFileToMatch);
                 if (!FileMatcher.IsMatch(filename, _filenamePattern))
                 {
                     return false;
                 }
 
-                var normalizedFileToMatch = FileUtilities.GetFullPathNoThrow(Path.Combine(_currentDirectory, fileToMatch));
                 return normalizedFileToMatch.StartsWith(_currentDirectory, StringComparison.OrdinalIgnoreCase);
```
Comment (Forgind) [inline src/Build/Utilities/FileSpecMatchTester.cs:84]: I'm kinda wondering if it's worth optimizing this to use EndsWith instead of StartsWith, as I imagine one of the most common differences is amd64 vs. no amd64. Probably not worth it.

File: src/Build/Evaluation/ItemSpec.cs (hunk 2 above)
Comment (Forgind) [reply src/Build/Evaluation/ItemSpec.cs:467]: This would make it hard to differentiate between "there were no referenced items" and "we couldn't enumerate them efficiently." Also, I don't think it's the critical case we're trying to optimize for here—I believe we're mostly looking at RemoveOperations without globs/wildcards.

Comment (Forgind) [reply src/Build/Evaluation/ItemSpec.cs:467]: As I understand it, we _do_ use this; we just haven't optimized for it. If it's easy to batch things, we do. If not, this returns null, and we do the slow thing. That shouldn't hurt perf too much because I don't know of cases when someone does a long series of RemoveOperations with wildcards, though I may be wrong on that point.

**Review by Forgind — COMMENTED** [review]
I don't see anything wrong with this, but it's late for me, so I don't want to approve.

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 3 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs]: Done, thank you!

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 4 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs]: Done, thank you!

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 7 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs]: Done, thank you!

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 6 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs]: Fixed by adding a struct Enumerator, thank you!

File: src/Build/Evaluation/ItemSpec.cs (hunk 1 above)
Comment (ladipro) [reply src/Build/Evaluation/ItemSpec.cs]: Wow, I was hoping C# would generate better code for an iterator function yield-returning a single item. I have applied your change. I find it less readable but it wins perf-wise, at least on Core.

File: src/Build/Evaluation/ItemSpec.cs (hunk 2 above)
Comment (ladipro) [reply src/Build/Evaluation/ItemSpec.cs:467]: I actually spent some time trying to come up with a combination of glob Includes and Removes that would be slow but couldn't find anything bad. It is orders of magnitude faster than the scenarios addressed with the PR.

This `null` is important, it tells the caller that rather than enumerating all referenced items and removing them from the running list, the caller should check each item in the running list and test it against the glob via `IsMatch`.

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 5 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs:509]: In the grand scheme of things since N is ~number of files on disk, I believe that O(logN) is fine. The alternative would be to use a regular dictionary with O(1) access and rebuilding it every time immutability is actually exercised. Which happens when the collection is added to the cache in `MemoizedOperation`. There may be optimizations that would remove the immutability requirement but I don't want to go there in this PR.

File: src/Build/Evaluation/ItemSpec.cs (hunk 2 above)
Comment (ladipro) [reply src/Build/Evaluation/ItemSpec.cs:467]: Yes, this is intentional. https://github.com/dotnet/msbuild/pull/6716#discussion_r681603275

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 9 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs:305]: The list is needed for cases where the same item (or, more precisely, an item with the same path-normalized value) is included multiple times.

```xml
  <ItemGroup>
    <A Include="file" />    
    <A Include="file" />    
  </ItemGroup>
```

When later we evaluate a remove that matches `file`, both items should be removed from the running list. To support immutability, we keep track of the collection version and clone the list when it would mutate an older version.

This shouldn't be common, only ~80 lists are created during a clean build of OC.

I went with a plain List for perf because this cloning is super rare (duplicate items + memoization) but you're right that `ImmutableList` is exactly what I need and simplifies the code a bit.

Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs]: Switched to `ImmutableList`, thank you!

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 3 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs:402]: I reverted this because it doesn't work well with the struct enumerator of `DictionaryValue`. A manual `foreach` avoids boxing and the expected number of items added is 1 so `UnionWith` wouldn't be helping even if it was optimized for multiple adds (which it looks like it is not).

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 4 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs:419]: I reverted this because it doesn't work well with the struct enumerator of `DictionaryValue`. A manual `foreach` avoids boxing and the expected number of items added is 1 so `UnionWith` wouldn't be helping even if it was optimized for multiple adds (which it looks like it is not).

File: src/Build/Evaluation/LazyItemEvaluator.cs [hunk 13]
```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
...
+                private void AddToDictionary(I item)
+                {
+                    string key = FileUtilities.NormalizePathForComparisonNoThrow(item.EvaluatedInclude, item.ProjectDirectory);
+
+                    if (!_dictionaryBuilder.TryGetValue(key, out var dictionaryValue))
+                    {
+                        dictionaryValue = new DictionaryValue(_version, item);
+                    }
+                    dictionaryValue.Add(_version, item);
+                    _dictionaryBuilder[key] = dictionaryValue;
```
[the remark this answers is not in the document]
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs:475]: Needs to be done in both cases - when adding a new `DictionaryValue` and when updating an existing one - because it's a struct.

File: src/Build/Utilities/FileSpecMatchTester.cs (hunk 12 above)
Comment (ladipro) [reply src/Build/Utilities/FileSpecMatchTester.cs:84]: I don't believe that `EndsWith` walks the string differently than `StartsWith`. They both use the same underlying comparison routines, just the (sub)strings being compared are different. We need the file path to start with the directory so `StartsWith` should be the right choice. My gut is that a manual implementation that walks the string backwards would be slower on average than the highly optimized BCL code and it is not worth it.

https://github.com/dotnet/runtime/blob/b3ab2eb4db830f6c520340f80651b806fd9dfce8/src/libraries/System.Private.CoreLib/src/System/Globalization/Ordinal.cs#L14

File: src/Build/Evaluation/LazyItemEvaluator.cs [hunk 14]
```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
...
+                {
+                    _version = version;
+                    _value = item;
+                }
+
+                public void Add(int version, I item)
+                {
+                    if (_value is List<I> list)
+                    {
+                        if (version != _version)
```
[the remark this answers is not in the document]
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs]: I removed the version variables and implemented immutability using `ImmutableList` as suggested in another comment.

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 11 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs]: Thank you, I've fixed the code to use `RemoveRange`.

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 8 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs]: Yes, done!

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 10 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs]: This is a great suggestion. I think I had the code structured differently but hit a technical issue which had me settle for this mess. I'll wait for CI to start working again to see where I stand with unit tests and will then tackle this. Thank you!

File: src/Build/Evaluation/ItemSpec.cs (hunk 2 above)
Comment (cdmihai) [reply src/Build/Evaluation/ItemSpec.cs:467]: > it tells the caller that rather than enumerating all referenced items and removing them from the running list, the caller should check each item in the running list and test it against the glob via IsMatch

Yup, which makes the calling code more complex (see https://github.com/dotnet/msbuild/pull/6716#discussion_r678762503)

File: src/Build/Utilities/FileSpecMatchTester.cs (hunk 12 above)
Comment (Forgind) [reply src/Build/Utilities/FileSpecMatchTester.cs:93]: Good point 👍

**Review by Forgind — APPROVED** [review]
I'm a little curious how few things you would have to remove for it to be not worth it to make this ordered collection. Probably not important, since it would be fast either way. Thanks!

File: src/Build/Evaluation/LazyItemEvaluator.OrderedItemDataCollection.cs [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+                            }
+                        }
+                    }
+
+                    // Finish by removing items from the list.
+                    if (keysToRemove != null)
+                    {
+                        dictionaryBuilder.RemoveRange(keysToRemove);
+                    }
+                    if (items != null)
```
[the remark this answers is not in the document]
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.OrderedItemDataCollection.cs]: Done, thank you.

File: src/Build/Evaluation/LazyItemEvaluator.OrderedItemDataCollection.cs [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+
+                    public I Current => (_value is IList<I> list) ? list[_index] : (I)_value;
+                    object System.Collections.IEnumerator.Current => Current;
+
+                    public void Dispose()
+                    { }
+
+                    public bool MoveNext()
+                    {
+                        int count = (_value is IList<I> list) ? list.Count : 1;
```
[the remark this answers is not in the document]
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.OrderedItemDataCollection.cs]: I have added the comment as suggested. Thank you!

**Comment by ladipro** [comment]
> I'm a little curious how few things you would have to remove for it to be not worth it to make this ordered collection. Probably not important, since it would be fast either way.

The remove operation is dominated by normalization of item values. This will run on the first remove either way. The extra overhead of building the dictionary should be relatively small and I'd be surprised if two removes were not already faster. I will measure and confirm, it's a fair question.

File: src/Build/Evaluation/LazyItemEvaluator.cs (hunk 10 above)
Comment (ladipro) [reply src/Build/Evaluation/LazyItemEvaluator.cs]: I have refactored it a bit as suggested and it is indeed nicer. For Update operation, it is possible that we can now revert the previous optimization (#5853) and achieve the same using the new datastructure but I'd like to try this separately later.

**Comment by ladipro** [comment]
There is a measurable regression in "small scale" removes, just as you suspected @Forgind. It slows down evaluation of a Hello World .NET Core project by a few percent. I will profile and make changes to address this.

**Comment by ladipro** [comment]
I've made the following changes:
- No longer maintaining the dictionary as immutable so it is forgotten when the list is added to the cache in `MemoizedOperation`. As @Therzok pointed out earlier, `ImmutableDictionary` has a cost and it turns out that we rarely take advantage of this memoization in real-world projects (when building OC I see <300 cache hits where we further modify the list as opposed to 20k hits with no modification).
- Enabled the new code path only for larger items lists (env variable configurable, >= 100 items by default) because the dictionary set-up work still showed in profiles for small ones.

I have updated the description with fresh measurements. There is a positive impact on evaluation time even if the dictionary is not used, due to other smaller optimizations, primarily caching the normalized value on `ItemData` and refactoring `FileSpecMatchTester` to prevent repeated normalizations.

File: src/Build/Evaluation/ItemDataCollectionValue.cs [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        public void Add(I item)
+        {
+            if (_value is not List<I> list)
+            {
+                list = new List<I>();
+                _value = list;
+            }
+            list.Add(item);
```
Comment (rainersigwald) [inline src/Build/Evaluation/ItemDataCollectionValue.cs:70]: Should we have some unit tests for this class? My first instinct is that this looks like it would drop a single item when adding the next one.

File: src/Build/Utilities/FileSpecMatchTester.cs [hunk 18]
```diff
@@ -41,31 +48,52 @@ public static FileSpecMatcherTester Parse(string currentDirectory, string fileSp
...
 
+        /// <summary>
+        /// Returns true if the given file matches this file spec.
+        /// </summary>
         public bool IsMatch(string fileToMatch)
         {
             Debug.Assert(!string.IsNullOrEmpty(fileToMatch));
 
+            // Historically we've used slightly different normalization logic depending on the type of matching
+            // performed in IsMatchNormalized. We have to keep doing it for compat.
```
Comment (rainersigwald) [inline src/Build/Utilities/FileSpecMatchTester.cs:59]: Can you give a bit more color on the types of scenarios where the difference is observable?

File: src/Build/Definition/Project.cs [hunk 19]
```diff
@@ -505,7 +507,7 @@ public static Project FromProjectRootElement(ProjectRootElement rootElement, Pro
                 options.SubToolsetVersion,
                 options.ProjectCollection ?? ProjectCollection.GlobalProjectCollection,
                 options.LoadSettings,
-                options.EvaluationContext);
+                options.EvaluationContext ?? _sharedEvaluationContext);
```
Comment (rainersigwald) [inline src/Build/Definition/Project.cs:510]: I'm not sure I understand the implications of this. Can you describe what it helps?

File: src/Build/Evaluation/ItemDataCollectionValue.cs (hunk 17 above)
Comment (ladipro) [reply src/Build/Evaluation/ItemDataCollectionValue.cs:70]: Good catch! We definitely don't have this covered with tests, will fix.

File: src/Build/Definition/Project.cs (hunk 19 above)
Comment (ladipro) [reply src/Build/Definition/Project.cs:510]: This was an experiment which was not supposed to become part of this PR. I'll remove it.

File: src/Build/Evaluation/ItemDataCollectionValue.cs (hunk 17 above)
Comment (ladipro) [reply src/Build/Evaluation/ItemDataCollectionValue.cs]: I have fixed the code, added a unit test, and made `ProjectItem_Tests` run with as well as without the directory-based optimization.

File: src/Build/Definition/Project.cs (hunk 19 above)
Comment (ladipro) [reply src/Build/Definition/Project.cs]: Removed.

File: src/Build/Utilities/FileSpecMatchTester.cs (hunk 18 above)
Comment (ladipro) [reply src/Build/Utilities/FileSpecMatchTester.cs:59]: Trying to unify the code to use the same path normalization function for pattern matching and for single path matching resulted in unit test failures for me. I can revisit if you want but generally I treat this as something that should not be changed just for code cleanliness, even if it affects only corner cases.

Comment (rainersigwald) [reply src/Build/Utilities/FileSpecMatchTester.cs:59]: I agree.

**Review by rainersigwald — APPROVED** [review]
